HW2/P2/parallel_vector.pyx

```diff
@@ -74,17 +74,35 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
        int idx, r
        omp_lock_t *locks = get_N_locks(counts.shape[0])
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):
+               # acquire locks, every element has its own lock
+               # to avoid dead-locking, 
+               # the locking order must be the same between different threads
+               if src[idx] > dest[idx]:
+                   acquire(&locks[src[idx]])
+                   acquire(&locks[dest[idx]])
+               elif src[idx] < dest[idx]:
+                   acquire(&locks[dest[idx]])
+                   acquire(&locks[src[idx]])
+               # avoid double locking
+               else:
+                   acquire(&locks[src[idx]])
+
                if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
                    counts[src[idx]] -= 1
+                   counts[dest[idx]] += 1
+
+               # release locks in the opposite order, to avoid dead-locking
+               if src[idx] > dest[idx]:
+                   release(&locks[dest[idx]])
+                   release(&locks[src[idx]])
+               elif src[idx] < dest[idx]:
+                   release(&locks[src[idx]])
+                   release(&locks[dest[idx]])
+               else:
+                   release(&locks[src[idx]])
```

Comment:
I did a very similar thing to find the minimum and ensure that we were locking/releasing in the right order. Someone mentioned on my pull request that the code could be shortened greatly by using min instead of doing all of the comparisons.